### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by ozgunozerk**
1. Turned operator overloads into member functions => now classes are more organized
2. Overall order or operator overloads are tidied (grouped as boolean, arithmetic, assignment)
3. Bug related to shift operation is fixed in uint128 and uint256 libraries

### CODE REVIEW

File: CUDA/encode-test.cu [hunk 1]
```diff
@@ -11,6 +11,8 @@ using namespace std;
 std::random_device dev;
 std::mt19937_64 rng(dev());
 
+#define num_piece 1024 * 32
```
Comment (nazar-pc) [inline CUDA/encode-test.cu]: This define doesn't seem to be used anywhere

File: CUDA/encode-test.cu [hunk 2]
```diff
@@ -25,68 +27,67 @@ uint256_t random_256()
...
-
-	a.high.high = std::bitset<64>(std::string("1111111111111111111111111111111111111111111111111111111111111111")).to_ullong();
-	a.high.low = std::bitset<64>(std::string("1111111111111111111111111111111111111111111111111111111111111111")).to_ullong();
-	a.low.high = std::bitset<64>(std::string("1111111111111111111111111111111111111111111111111111111111111111")).to_ullong();
-	a.low.low = std::bitset<64>(std::string("1111111111111111111111111111111111111111111111111111111100111100")).to_ullong();
 
-	cudaMalloc(&d_a, 4 * (sizeof(unsigned long long)));
-	cudaMemcpy(d_a, &a, 4 * (sizeof(unsigned long long)), cudaMemcpyHostToDevice);
+	uint256_t* piece, * d_piece, * d_nonce;
+	cudaMallocHost(&piece, 4 * sizeof(unsigned long long) * 128);
```
Comment (nazar-pc) [inline CUDA/encode-test.cu]: I would probably save `4 * sizeof(unsigned long long) * 128` into some named constant

File: CUDA/encode-test.cu [hunk 3]
```diff
@@ -25,68 +27,67 @@ uint256_t random_256()
...
 
-	//cout << "a: " << bitset<64>(a.high.high) << bitset<64>(a.high.low) << bitset<64>(a.low.high) << bitset<64>(a.low.low) << endl;
+	unsigned char* byte_piece = (unsigned char*)piece;
+	for (int i = 0; i < 32 * 128; i++)
```
Comment (nazar-pc) [inline CUDA/encode-test.cu]: Here again `32 * 128` feels like a random choice unless you know where those numbers are coming from.

File: CUDA/encode-test.cu (hunk 1 above)
Comment (ozgunozerk) [reply CUDA/encode-test.cu]: true, leftover from parallel version, deleting now

File: CUDA/encode-test.cu (hunk 2 above)
Comment (ozgunozerk) [reply CUDA/encode-test.cu]: Ok

File: CUDA/encode-test.cu (hunk 3 above)
Comment (ozgunozerk) [reply CUDA/encode-test.cu]: 32 * 128 => 4096
so we treat each integer as a byte (to simulate the same behaviour in Rust code). 
This choice of numbers 32*128 was not random. Piece has 4096 bytes, Expanded_iv has 32 bytes.

Comment (nazar-pc) [reply CUDA/encode-test.cu]: I know where those numbers are coming from, the comment was generally about the fact that those are called "magic numbers" and should probably be named constants or have some comment explaining where those specific values are coming from.

Comment (ozgunozerk) [reply CUDA/encode-test.cu]: I see your point, commented them already :)

**Review by nazar-pc — APPROVED** [review]

**Pull request merged**